dev_scripts/figures.py

```diff
@@ -1,6 +1,43 @@
 import matplotlib.pyplot as plt
+import matplotlib.colors as mcolors
 from collections import defaultdict
 from reporting.generate import get_summary_data
+from runner.master_log import MasterLog
+from utils.constants import DATA_DIR, EventType
+from utils.llm import tokens_in_text
+from utils.math import mean_std
+import numpy as np
+
+
+def message_tokens(message: str, agt_name: str) -> int:
+    model = agt_name.split(" - ")[1]
+    if model.startswith("gpt-4"):
+        model = "gpt-4-turbo-2024-04-09"
+    elif model.startswith("claude-3"):
+        model = "claude-3-opus-20240229"
+    else:
+        model = "together_ai/meta-llama-Llama-3-70b-chat-hf"
+    return tokens_in_text(message, model)
+
+
+def speed_stats(run: str, agt_name: str) -> tuple[float, float]:
+    log = MasterLog(DATA_DIR.joinpath("tests", run, "results", agt_name, "master_log.jsonl"))
+    log.load()
+    speeds = list()
+    total_tokens = 0
+    total_time = 0
+    for evt in log.test_events(event_type={EventType.SEND_MESSAGE, EventType.RESPONSE_MESSAGE}):
+        if evt.type == EventType.SEND_MESSAGE:
+            t0 = evt.timestamp
+        else:
+            msg = evt.data["message"]
+            t = (evt.timestamp - t0).total_seconds()
+            tokens = message_tokens(msg, agt_name)
+            speeds.append(tokens / t)
+            total_tokens += tokens
+            total_time += t
+    # return mean_std(speeds)
+    return total_tokens / total_time, 0
```

Comment:
So we are counting response times for the test messages only?

Reply:
Ah yes... that was me trying to get a reliable speed measurement. LLMs don't actually reply to filler messages, so this is how I managed to skip them. I could still consider the fillers for LTM agents though. However, I am still not happy with the results... I will try a couple of things more today and send you some previews.